src/V3PerPackage/StringLocker.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace NuGet.Services.V3PerPackage
+{
+    public class StringLocker
+    {
+        private readonly object _semaphoresLock = new object();
+        private readonly Dictionary<string, SemaphoreSlim> _semaphores = new Dictionary<string, SemaphoreSlim>();
+
+        private readonly ILogger<StringLocker> _logger;
+
+        public StringLocker(ILogger<StringLocker> logger)
+        {
+            _logger = logger;
+        }
+
+        public async Task<IDisposable> AcquireAsync(string semaphoreName)
+        {
+            SemaphoreSlim semaphore;
+            lock(_semaphores)
+            {
+                if (!_semaphores.TryGetValue(semaphoreName, out semaphore))
+                {
+                    semaphore = new SemaphoreSlim(1);
+                    _semaphores[semaphoreName] = semaphore;
+                }
+            }
+
+            var acquired = await semaphore.WaitAsync(TimeSpan.Zero);
+            if (!acquired)
+            {
+                _logger.LogInformation("The string {SemaphoreName} has already been acquired. Waiting.", semaphoreName);
+                await semaphore.WaitAsync();
+            }
+            
+            return new DisposableSemaphoreSlim(semaphore);
+        }
+
+        public async Task<IDisposable> AcquireAsync(IEnumerable<string> semaphoreNames)
```

Comment:
>AcquireAsync [](start = 39, length = 12)

I would add a quick comment that this API should be used to avoid deadlocks

Reply:
Sure

Reply:
Fixed.